src/main/java/tools/jackson/databind/introspect/AnnotatedCreatorCollector.java

```diff
@@ -66,24 +66,15 @@ Creators collect(Class<?> primaryMixIn)
          * ignorable after all annotations have been properly collapsed.
          */
         // AnnotationIntrospector is null if annotations not enabled; if so, can skip:
-        if (_collectAnnotations) {
-            if (_defaultConstructor != null) {
-                if (_intr.hasIgnoreMarker(_config, _defaultConstructor)) {
-                    _defaultConstructor = null;
-                }
-            }
-            // count down to allow safe removal
-            for (int i = constructors.size(); --i >= 0; ) {
-                if (_intr.hasIgnoreMarker(_config, constructors.get(i))) {
-                    constructors.remove(i);
-                }
-            }
-            for (int i = factories.size(); --i >= 0; ) {
-                if (_intr.hasIgnoreMarker(_config, factories.get(i))) {
-                    factories.remove(i);
-                }
-            }
+        if (_collectAnnotations && _defaultConstructor != null && _intr.hasIgnoreMarker(_config, _defaultConstructor)) {
+            _defaultConstructor = null;
         }
+
+        constructors.removeIf(constructor ->
+                constructor == null || (_collectAnnotations && _intr.hasIgnoreMarker(_config, constructor)));
```

Comment:
Is null check needed? It wasn't there earlier?

Reply:
It's now necessary because access checks have been pushed up from the point of use (when constructor is called) to earlier in deserializer creation (when constructors and factories are collected) so they are now filtered out earlier too. I added a comment.